### PR TITLE
Correction of the work with comments and general improvements

### PR DESCRIPTION
**Pull request by Ulyanov-programmer**
List of changes:

- Fixed a bug where rules with comments without properties could lose comments embedded in them.
- The code that was duplicated several times was deleted and moved to a new function.
- General code improvements: `foreach` has been replaced with `for`, and some methods have been optimized.

**All tests were passed correctly.**

### CODE REVIEW

File: index.test.js [hunk 1]
```diff
@@ -605,17 +605,21 @@ test('clears empty selector after comma', () => {
 })
 
 test('moves comment with rule', () => {
-  run('a { /*B*/ /*B2*/ b {} }', '/*B*/ /*B2*/ a b {}')
+  run('a { /*B*/ /*B2*/ b {} }', 'a { /*B*/ /*B2*/ } a b {}')
```
Comment (ai) [inline index.test.js:608]: Is it a expected output?

```css
.block {
  /* Element comment */
  .element {
  }
}
```

will not be moved too?

Comment (Ulyanov-programmer) [reply index.test.js:608]: I didn't understand you.
Your code should be converted to:
```css
.block {
  /* Element comment */
}
.block .element {
}
```
It was from this logic that I transformed the test.

Comment (ai) [reply index.test.js:608]: > Your code should be converted to

Is it expected output?

Seems like very soon other developers will come and will open new issues asking to not move this kind of comments

Comment (Ulyanov-programmer) [reply index.test.js:608]: Do you really think that the comments embedded in the rules should remain outside?
If a comment is embedded in a rule, it should remain in it. If a comment is not embedded in a rule, it must be outside of it. It's logical, isn't it?

Comment (Ulyanov-programmer) [reply index.test.js:608]: In fact, the plugin originally assumed this logic - when I wrote a comment in the rule, it remained in it (see the original [issue](https://github.com/postcss/postcss-nested/issues/162)). But in one of the cases there was a mistake that I fixed.

Comment (ai) [reply index.test.js:608]: > If a comment is embedded in a rule, it should remain in it

It is our point of view, but it is just only one of many.

Other people think that comments and bound to the rules which is bellow. For many people `/* Element comment */` here should be bound to `.element` (because this comment explain something about the `.element`).

```css
.block {
  /* Element comment */
  .element {
  }
}
```

Look more realistic example:

```css
.nav {
  /* Home button */
  &_logo {
    width: 100px;
    height: 50px;
  }
}
```

To solve this issue we need to add many different examples to the tests for many different edge-cases. This is why I do not want to do this task, since it is not a simple fix but a big research.

Comment (Ulyanov-programmer) [reply index.test.js:608]: > To solve this issue we need to add many different examples to the tests for many different edge-cases

What other extreme cases, I wonder?
The joke is not funny. This is solved by one optional parameter in the API. If OFF, comments are saved outside the rules, if ON, inside.

Comment (ai) [reply index.test.js:608]: > What other extreme cases, I wonder?

This is part of research.

> This is solved by one optional parameter 

Adding option is the last way.

Can you explain your business case? Why exactly do you need to keep the comment inside the rule. Maybe we can add some detection based on the comment content on new lines around it, or next child

Comment (ai) [reply index.test.js:608]: Got it. The problem is that it is unique feature for mostly 1-2 people. I do not want to change behvaiour for everyone.

The suggestions:
1. You write a plugins which will run before (to hide comment to `Rule.props`) and after `postcss-nested` (to restore them).
2. We can add `keepComment` option to the `postcss-nested` but only if it will be very small to not reduce code maintaibility.

Comment (Ulyanov-programmer) [reply index.test.js:608]: There is another option: most often _(a matter of formatting_) there is an empty space between the properties and the nested rule. If they want to describe a rule with a comment, the comment is "pressed" to the rule from above, leaving no space.

This can be used to separate comments into those that cannot be left inside the rule and those that can.

Example:
```css
p {
  /* comment inside the rule, save as is */

  /* comment for the rule below, move with the rule */
  a {
  }
}
```

Comment (ai) [reply index.test.js:608]: Yes, let's use empty new line to detect it

Comment (Ulyanov-programmer) [reply index.test.js:608]: Okay, I'll take care of it in the next few days and let you know when the result is.

**Comment by Ulyanov-programmer** [comment]
Changed the code. Now it handles CSS as indicated in my example just above.

**Comment by Ulyanov-programmer** [comment]
Well, or something like that. I didn't seem to miss anything.

**Comment by ai** [comment]
Hm, it may be an ESLint issue since we should support `>=18`

**Comment by ai** [comment]
I will look

File: index.test.js [hunk 2]
```diff
@@ -604,16 +604,37 @@ test('clears empty selector after comma', () => {
...
 
-test('moves comment with rule', () => {
-  run('a { /*B*/ /*B2*/ b {} }', '/*B*/ /*B2*/ a b {}')
+test("Save the parent's comment", () => {
+  run('a { /*i*/ b {} }', 'a { /*i*/ } a b {}')
 })
 
-test('moves comment with at-rule', () => {
-  run('a { /*B*/ @media { one: 1 } }', '/*B*/ @media {a { one: 1 } }')
+test('Save the comments for the parent and child', () => {
```
Comment (ai) [inline index.test.js:611]: Can we add test with `\n` like:

```css
a { 
    /*i*/ 

    b {} }
```

Seems like often you will not have inside-commend.

Comment (Ulyanov-programmer) [reply index.test.js:611]: We can add it. I've already checked, it will work properly.

Comment (ai) [reply index.test.js:611]: Yes, let's add it to not break in the future

**Comment by ai** [comment]
I removed Node.js 12-16 support from `main`. No need to rebase.

**Comment by Ulyanov-programmer** [comment]
Haven't heard from you for a long time, have you been waiting for me to add a test?

**Comment by ai** [comment]
> Haven't heard from you for a long time, have you been waiting for me to add a test?

Yes

**Comment by Ulyanov-programmer** [comment]
> > Haven't heard from you for a long time, have you been waiting for me to add a test?
> 
> Yes

No comment, it's fabulous.

**Comment by ai** [comment]
Thanks. Released in 7.0.
